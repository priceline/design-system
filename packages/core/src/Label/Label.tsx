import React from 'react'
import styled from 'styled-components'
import {
  FontSizeProps,
  FontWeightProps,
  LineHeightProps,
  SpaceProps,
  TextStyleProps,
  WidthProps,
  compose,
  fontSize,
  fontWeight,
  letterSpacing,
  lineHeight,
  space,
  textStyle,
  width,
} from 'styled-system'
import { typographyAttrs } from '../utils/attrs/typographyAttrs'
import { applyVariations, getPaletteColor } from '../utils/utils'

const nowrap = (props) =>
  props.nowrap
    ? {
        whiteSpace: 'nowrap',
      }
    : null

const accessiblyHide = (props) =>
  props.hidden
    ? {
        position: 'absolute',
        width: '1px',
        height: '1px',
        clip: 'rect(1px, 1px, 1px, 1px)',
      }
    : null

const truncate = (props) =>
  props.truncate
    ? {
        whiteSpace: 'nowrap',
        overflow: 'hidden',
        textOverflow: 'ellipsis',
      }
    : null

/**
 * @public
 */
export type LabelProps = SpaceProps &
  FontSizeProps &
  FontWeightProps &
  LineHeightProps &
  TextStyleProps &
  WidthProps &
  Partial<Omit<HTMLLabelElement, 'children'>> & {
    children?: React.ReactNode
    color?: string
    autoHide?: boolean
    nowrap?: boolean
    for?: string
    as?: unknown
    truncate?: boolean
    onClick?: (evt: unknown) => void
  }

/**
 * @public
 */
export const Label: React.FC<LabelProps> & { isLabel?: boolean } = styled.label.attrs((props) => ({
  ...typographyAttrs(props),
  ...props,
}))`
  display: block;
  width: 100%;
  margin: 0;
  z-index: 1;
  color: ${getPaletteColor('base')};
  ${(props) => (props.bg ? `background-color: ${getPaletteColor(props.bg, 'base')(props)};` : '')}
  ${(props) => (props.onClick ? 'cursor: pointer;' : '')}

  ${applyVariations('Label')}

  ${nowrap}
  ${accessiblyHide}
  ${truncate}

  ${(props) => compose(fontSize, fontWeight, lineHeight, letterSpacing, space, textStyle, width)(props)}
`

Label.defaultProps = {
  color: 'text.light',
  textStyle: 'label',
}

Label.displayName = 'Label'
Label.isLabel = true
