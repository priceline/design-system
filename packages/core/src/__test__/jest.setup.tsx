import '@testing-library/jest-dom'
import 'jest-styled-components'

global.requestAnimationFrame =
  global.requestAnimationFrame ||
  function _raf(cb) {
    return setTimeout(cb, 0)
  }
