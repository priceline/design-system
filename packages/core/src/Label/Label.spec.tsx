import { fireEvent, render, screen } from '../__test__/testing-library'
import { Label } from './Label'

describe('Label', () => {
  test('it renders', () => {
    const { asFragment } = render(<Label />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('Label hidden renders', () => {
    render(<Label hidden>Label Children</Label>)

    expect(screen.getByText('Label Children')).toHaveStyleRule('position', 'absolute')
    expect(screen.getByText('Label Children')).toHaveStyleRule('width', '1px')
    expect(screen.getByText('Label Children')).toHaveStyleRule('height', '1px')
    expect(screen.getByText('Label Children')).toHaveStyleRule('clip', 'rect(1px,1px,1px,1px)')
  })

  test('Label nowrap renders', () => {
    render(<Label nowrap>Label Children</Label>)

    expect(screen.getByText('Label Children')).toHaveStyleRule('white-space', 'nowrap')
  })

  test('Label truncate renders', () => {
    render(<Label truncate>Label Children</Label>)

    expect(screen.getByText('Label Children')).toHaveStyleRule('white-space', 'nowrap')
    expect(screen.getByText('Label Children')).toHaveStyleRule('overflow', 'hidden')
    expect(screen.getByText('Label Children')).toHaveStyleRule('text-overflow', 'ellipsis')
  })

  test('Label width renders', () => {
    render(
      <Label width={1 / 2} nowrap>
        Label Children
      </Label>
    )

    expect(screen.getByText('Label Children')).toHaveStyleRule('white-space', 'nowrap')
    expect(screen.getByText('Label Children')).toHaveStyleRule('width', '50%')
  })

  test('Label clickable renders with cursor pointer', () => {
    const mockOnClick = jest.fn()
    render(<Label onClick={mockOnClick}>Clickable Label</Label>)

    const label = screen.getByText('Clickable Label')
    expect(label).toHaveStyleRule('cursor', 'pointer')

    expect(mockOnClick).toHaveBeenCalledTimes(0)
    fireEvent.click(label)
    expect(mockOnClick).toHaveBeenCalledTimes(1)
  })
})
