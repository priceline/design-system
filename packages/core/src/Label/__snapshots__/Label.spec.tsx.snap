// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label it renders 1`] = `
<DocumentFragment>
  .c0 {
  font-family: 'Montserrat','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.4;
  font-weight: 500;
}

.c0 * {
  box-sizing: border-box;
}

.c1 {
  display: block;
  width: 100%;
  margin: 0;
  z-index: 1;
  color: #4f6f8f;
  font-weight: 700;
  font-size: 10px;
  line-height: 1.4;
  -webkit-letter-spacing: 0.2px;
  -moz-letter-spacing: 0.2px;
  -ms-letter-spacing: 0.2px;
  letter-spacing: 0.2px;
}

<div
    class="c0"
  >
    <label
      class="c1"
      color="text.light"
      font-size="10px"
      font-weight="700"
      letter-spacing="0.2px"
    />
  </div>
</DocumentFragment>
`;
